### PR TITLE
[FEATURE] Better protein ID support for mzTabExporter

### PR DESCRIPTION
**Pull request by jpfeuffer**
mzTabExporter now writes multiple PeptideEvidences of a PeptideID in multiple PSM rows as suggested by the format. Enables reporting of shared peptides.
Also adds meta value support for PRT level. Currently adds a column for every occurring meta value.

### CODE REVIEW

File: src/topp/MzTabExporter.cpp [hunk 1]
```diff
@@ -120,6 +120,83 @@ namespace OpenMS
...
+          }
+
+          row.accession = MzTabString(peptide_evidences[i].getProteinAccession());
+
+          rows.push_back(row);
+        }
+      }
+      else
+      { // report without pep evidence information
+        row.pre = MzTabString("-");
```
Comment (timosachsenberg) [inline src/topp/MzTabExporter.cpp]: I don't know this by heat but did you check if "-" or "null" is correct?

File: src/topp/MzTabExporter.cpp [hunk 2]
```diff
@@ -437,38 +514,87 @@ namespace OpenMS
           const std::vector<ProteinIdentification::ProteinGroup> indist_groups = it->getIndistinguishableProteins();
           const std::vector<ProteinHit> protein_hits = it->getHits();
 
+          // pre-analyze data for occuring meta values at protein hit level
+          // these are used to build optional columns containing the meta values in internal data structures
+          set<String> protein_hit_user_value_keys;
+          for (Size i = 0; i < protein_hits.size(); ++i)
```
Comment (timosachsenberg) [inline src/topp/MzTabExporter.cpp]: looks a bit like the helper code chris wrote. would it be possible to reuse it or move your code there?

File: src/topp/MzTabExporter.cpp [hunk 3]
```diff
@@ -437,38 +514,87 @@ namespace OpenMS
...
-//          std::vector<MzTabOptionalColumnEntry> opt_; // Optional Columns must start with “opt_”
-            // optional column
+         // std::vector<MzTabOptionalColumnEntry> opt_; // Optional Columns must start with “opt_”
+
+            // create opt_ columns for protein hit user values
+            for (set<String>::const_iterator mit = protein_hit_user_value_keys.begin(); mit != protein_hit_user_value_keys.end(); ++mit)
+            {
+              MzTabOptionalColumnEntry opt_entry;
+              const String& key = *mit;
+              opt_entry.first = String("opt_prt_") + key;
```
Comment (timosachsenberg) [inline src/topp/MzTabExporter.cpp]: the prefix is afaik not correct: "assay, study variable, MS run or “global”"
probably opt_global would fit best

File: src/topp/MzTabExporter.cpp (hunk 1 above)
Comment (timosachsenberg) [inline src/topp/MzTabExporter.cpp]: just checked it:
Amino acid preceding the peptide (coming from the PSM) in the protein
sequence. If unknown “null” MUST be used, if the peptide is N-terminal “-“
MUST be used. 
So - is wrong

File: src/topp/MzTabExporter.cpp (hunk 2 above)
Comment (jpfeuffer) [inline src/topp/MzTabExporter.cpp]: Yep. Works with that. Do you see a possibility to use that code in the other methods for FeatXML and ConsXML? Problem is the two level hierarchy of Features and Identifications in there.

Comment (timosachsenberg) [inline src/topp/MzTabExporter.cpp]: For Feature it should work but for the lower level PeptideIdentification I think either needs to define custom iterators or use  something like http://www.boost.org/doc/libs/1_58_0/libs/iterator/doc/iterator_adaptor.html to e.g. generate a "deep iterator" that iterates over all meta values of the underlying PeptideIdentification objects of each Feature. But maybe there are other / better ways. C++ guru to the rescue?

File: src/topp/MzTabExporter.cpp (hunk 3 above)
Comment (jpfeuffer) [inline src/topp/MzTabExporter.cpp]: Okay, according to the standard you are right. I was adapting to previous code in the class. I guess I have to change them, too.

Comment (jpfeuffer) [inline src/topp/MzTabExporter.cpp]: What happens when metainfos on different levels (e.g. Feature and PeptideID) have the same name?

Comment (timosachsenberg) [inline src/topp/MzTabExporter.cpp]: For sublevel metavalues I think we could add _psm_ etc. For the top level one I would not add anything.

**Comment by timosachsenberg** [comment]
+1

**Pull request merged**